File: Recursion/src/main/java/Solution78.java
import java.util.*;

public class Solution78 {
    public List<List<Integer>> subsets(int[] nums) {
        recur(0, nums);

        return ans;

    }

    List<List<Integer>> ans = new ArrayList<>();
    List<Integer> subAns = new ArrayList<>();

    private void recur(int i, int[] nums) {
        if (i == nums.length) {
            // 需要创建一个新数组
            ans.add(new ArrayList<>(subAns));
            return;
        }

        recur(i + 1, nums);
        subAns.add(nums[i]);
        recur(i + 1, nums);

        subAns.remove(subAns.size() - 1);

    }

    public static void main(String[] args) {
        Solution78 s = new Solution78();
        int[] nums = { 1, 2, 3 };
        List<List<Integer>> res = s.subsets(nums);
        System.out.println(res);
    }
}

File: Recursion/src/main/python/solution78.py
# https://leetcode-cn.com/problems/subsets/

from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.ans = []
        self.sub_ans = []
        self.recur(0, nums)

        return self.ans

    def recur(self, n, nums):
        if n == len(nums):
            # 需要做个切片
            self.ans.append(self.sub_ans[:])
            return

        self.recur(n + 1, nums)

        self.sub_ans.append(nums[n])

        self.recur(n + 1, nums)

        self.sub_ans.pop()


s = Solution()
nums = [1, 2, 3]
# nums = [0]
res = s.subsets(nums)

print(res)